drive/applications/application_serial_commands.cpp

```
@@ -0,0 +1,299 @@
+  }
+
+  bool prefix_match(char const* prefix, std::span<hal::byte> str)
+  {
+    for (size_t pi = 0;; pi++) {
+      char target_char = prefix[pi];
+      bool given_end = pi >= str.size();
+      bool target_end = target_char == '\0';
+      if (given_end || target_end) {
+        return given_end && target_end;
+      }
+      if (str[pi] != target_char) {
+        return false;
+      }
+    }
+    return false;
+  }
+
+public:
+  // handle() reads the currently available bytes from the serial console and
+  // executes a command if it is complete.
+  void handle(hal_serial& console, std::span<command_def> commands)
+  {
+    while (true) {
+      switch (readline(console)) {
+        case 0:
+          break;
+        case 1:
+          goto stop_reading;
+        case 2:
+        case 3:
+          return;
+      }
+    }
+
+  stop_reading:
+
+    // size = maximum # of segments
+    //
+    // supposing the prefix is 1 char and each param is 1 char -> 1/2 of the
+    // line must be spaces.
+    //
+    // thus, max # of segments = 256/2 = 128
+    std::array<std::span<hal::byte>, 128> segments;
+    std::span<std::span<hal::byte>> segview{ segments };
+    parse(segview);
+    if (segview.size() == 0) {
+      return;
+    }
+
+    std::span<hal::byte> prefix = segments[0];
+    std::span<std::span<hal::byte>> params{ segments.begin() + 1,
+                                            segview.size() - 1 };
+
+    for (size_t i = 0; i < commands.size(); i++) {
+      command_def cmd = commands[i];
+      if (!prefix_match(cmd.prefix, prefix)) {
+        continue;
+      }
+
+      try {
+        cmd.callback(params);
+      } catch (hal::exception e) {
+        switch (e.error_code()) {
+          case std::errc::argument_out_of_domain:
+            hal::print(*console, "Error: invalid argument length or type\n");
+            break;
+          default:
+            hal::print<32>(*console, "Error code: %d\n", e.error_code());
+            break;
```

I think you should throw the exception incase the application has some from of error handling it can preform, or if it is fatal end the program rather than ignoring what ever got thrown.